jbmc/src/java_bytecode/assignments_from_json.cpp

```diff
@@ -0,0 +1,828 @@
+{
+  if(!json.is_object())
+    return false;
+  const auto &json_object = to_json_object(json);
+  if(json_object.find("@nondetLength") != json_object.end())
+    return (json["@nondetLength"].is_true());
+  return false;
+}
+
+/// For typed versions primitive, string or array types, looks up their untyped
+/// contents with the key specific to their type.
+static jsont get_untyped(const jsont &json, const std::string &object_key)
+{
+  if(has_type(json) || has_nondet_length(json))
+    return json[object_key];
+  return json;
+}
+
+/// \ref get_untyped for primitive types.
+static jsont get_untyped_primitive(const jsont &json)
+{
+  return get_untyped(json, "value");
+}
+
+/// \ref get_untyped for array types.
+/// char arrays are treated as a special case as they are represented as an
+/// array of a single String element by json-io, rather than an array of one or
+/// more char elements.
+static json_arrayt
+get_untyped_array(const jsont &json, const typet &element_type)
+{
+  const jsont untyped = get_untyped(json, "@items");
+  PRECONDITION(untyped.is_array());
+  const auto &json_array = to_json_array(untyped);
+  if(element_type == java_char_type())
+  {
+    PRECONDITION(json_array.size() == 1);
+    const auto &first = *json_array.begin();
+    PRECONDITION(first.is_string());
+    const auto &json_string = to_json_string(first);
+
+    const auto wide_string = utf8_to_utf16_native_endian(json_string.value);
+    auto string_range = make_range(wide_string.begin(), wide_string.end());
+    const auto json_range = string_range.map([](const wchar_t &c) {
+      const std::u16string u16(1, c);
+      return json_stringt{utf16_native_endian_to_utf8(u16)};
+    });
+    return json_arrayt{json_range.begin(), json_range.end()};
+  }
+  return json_array;
+}
+
+/// \ref get_untyped for string types.
+/// Note that this differs from the standard serialization of java.lang.String
+/// in json-io, but is consistent with the serialization of StringBuilder and
+/// StringBuffer.
+static jsont get_untyped_string(const jsont &json)
+{
+  return get_untyped(json, "value");
+}
+
+/// Given a JSON representation of a (non-array) reference-typed object and a
+/// type inferred from the type of a containing array, get the runtime type of
+/// the corresponding pointer expression.
+/// \param json: JSON representation of a non-array object. If it contains a
+///   `@type` field, this takes priority over \p type_from_array. Types for non-
+///   array objects are stored in the JSON in the format
+///   "my.package.name.ClassName".
+/// \param type_from_array: may contain an element type name given by a
+///   containing array. Such types are stored in the form
+///   "Lmy.package.name.ClassName;".
+/// \param symbol_table: used to look up the type given its name.
+/// \return runtime type of the object, if specified by at least one of the
+///   parameters.
+static optionalt<java_class_typet> given_runtime_type(
```

Comment:
Because of how I'm calling it:
```
      const auto runtime_type =
        given_runtime_type(json, type_from_array, info.symbol_table))
```
I can't call both the variable and the function `runtime_type`.

Reply:
It's not ideal because it feels that `runtime_type` and `given_runtime_type` are two different things. What about `runtime_type_from_json`?

Reply:
I renamed the function to `runtime_type` and I'm now calling it like
```
      const auto runtime_type =
        ::runtime_type(json, type_from_array, info.symbol_table))
```
looking up `runtime_type` in the global namespace. Looks like this will do the trick (thanks @jeannielynnmoulton ).